File: apidemo/voicedemo/src/main/java/com/gdkdemo/voice/VoiceDemoActivity.java
package com.gdkdemo.voice;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.view.MotionEvent;

import com.gdkdemo.voice.common.VoiceDemoConstants;
import com.google.android.glass.touchpad.Gesture;
import com.google.android.glass.touchpad.GestureDetector;

import java.util.ArrayList;


// The "main" activity...
public class VoiceDemoActivity extends Activity
{
    // For tap events
    private GestureDetector mGestureDetector;
    // Voice action.
    private String voiceAction = null;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        Log.d("onCreate() called.");

        setContentView(R.layout.activity_voicedemo);

        // For gesture handling.
        mGestureDetector = createGestureDetector(this);

        voiceAction = getVoiceAction(getIntent());
        if(Log.I) Log.i("voiceAction = " + voiceAction);
        processVoiceAction(voiceAction);
    }

    @Override
    protected void onResume()
    {
        super.onResume();
        Log.d("onResume() called.");

        // This makes it impossible to exit the app.
//        voiceAction = getVoiceAction(getIntent());
//        if(Log.I) Log.i("voiceAction = " + voiceAction);
//        processVoiceAction(voiceAction);
    }

    // Returns the "first" word from the phrase following the prompt.
    private String getVoiceAction(Intent intent)
    {
        if(intent == null) {
            return null;
        }
        String action = null;
        Bundle extras = intent.getExtras();
        ArrayList<String> voiceActions = null;
        if(extras != null) {
            voiceActions = extras.getStringArrayList(RecognizerIntent.EXTRA_RESULTS);
            if(voiceActions != null && !voiceActions.isEmpty()) {
                if(Log.D) {
                    for(String a : voiceActions) {
                        Log.d("action = " + a);
                    }
                }
                action = voiceActions.get(0);
            }
        }
        return action;
    }

    // Opens the WordDictation activity,
    // or, quits the program.
    private void processVoiceAction(String voiceAction)
    {
        if(voiceAction != null) {
            if(voiceAction.equals(VoiceDemoConstants.ACTION_START_DICTATION)) {
                openVoiceDictationActivity();
            } else if(voiceAction.equals(VoiceDemoConstants.ACTION_STOP_VOICEDEMO)) {
                Log.i("VoiceDemo activity has been terminated upon start.");
                this.finish();
            } else {
                Log.w("Unknown voice action: " + voiceAction);
            }
        } else {
            Log.w("No voice action provided.");
        }
    }

    private void openVoiceDictationActivity()
    {
        Intent intent = new Intent(this, VoiceDictationActivity.class);
        startActivity(intent);
    }


    ///////////////////////////////////////////////////////
    /// Gesture handling
    //

    @Override
    public boolean onGenericMotionEvent(MotionEvent event)
    {
        if (mGestureDetector != null) {
            return mGestureDetector.onMotionEvent(event);
        }
        return super.onGenericMotionEvent(event);
    }

    private GestureDetector createGestureDetector(Context context)
    {
        GestureDetector gestureDetector = new GestureDetector(context);
        //Create a base listener for generic gestures
        gestureDetector.setBaseListener( new GestureDetector.BaseListener() {
            @Override
            public boolean onGesture(Gesture gesture) {
                if(Log.I) Log.i("gesture = " + gesture);
                if (gesture == Gesture.TAP) {
                    openVoiceDictationActivity();
                    return true;
                } // etc...
                return false;
            }
        });
        return gestureDetector;
    }

}

File: xamarindemo/VoiceDemo/VoiceDemoActivity.cs

/*
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.speech.RecognizerIntent;
import android.view.MotionEvent;

import com.gdkdemo.voice.common.VoiceDemoConstants;
import com.google.android.glass.touchpad.Gesture;
import com.google.android.glass.touchpad.GestureDetector;

import java.util.ArrayList;
*/
using System;
using System.Linq;
using Android.Glass.Touchpad;
using Android.App;
using Android.OS;
using Android.Speech;
using Android.Content;
using VoiceDemo.Common;
using Android.Views;
using System.Collections.Generic;
using Android.Gestures;



namespace VoiceDemo
{

	[Activity (Label = "VoiceDemo", MainLauncher = true)]
	[IntentFilter(new[] {"com.google.android.glass.action.VOICE_TRIGGER"})]
	public class VoiceDemoActivity : Activity
	{
	    // For tap events
		private Android.Glass.Touchpad.GestureDetector _gestureDetector;
	    // Voice action.
		private string voiceAction = null;

		protected override void OnCreate(Bundle savedInstanceState)
	    {
			base.OnCreate(savedInstanceState);
			//Log.d("onCreate() called.");

			SetContentView(Resource.Layout.activity_voicedemo);

	        // For gesture handling.
			_gestureDetector = CreateGestureDetector(this);

			voiceAction = GetVoiceAction(Intent);
			//if(Log.I) Log.Info("voiceAction = " + voiceAction);
			ProcessVoiceAction(voiceAction);
	    }

		protected override void OnResume()
	    {
			base.OnResume();
			//Log.d("onResume() called.");

	        // This makes it impossible to exit the app.
	//        voiceAction = getVoiceAction(getIntent());
	//        if(Log.I) Log.i("voiceAction = " + voiceAction);
	//        ProcessVoiceAction(voiceAction);
	    }

	    // Returns the "first" word from the phrase following the prompt.
		private string GetVoiceAction(Intent intent)
	    {
	        if(intent == null) 
			{
	            return null;
	        }
			string action = null;
			Bundle extras = intent.GetBundleExtra("Extras");// GetStringArrayListExtra(RecognizerIntent.ExtraResults);


			IList<string> voiceActions = null;
	        if(extras != null) 
			{
				voiceActions = extras.GetStringArrayList(RecognizerIntent.ExtraResults);

				if(voiceActions != null && !voiceActions.Any()) 
				{
					//if(Log.D) 
					//{
						foreach(string a in voiceActions) 
						{
							//Log.Debug("action = " + a);
						//Trace ("Action = ", a);
	                    }
					//}
					action = voiceActions.ElementAt(0);
	            }
	        }
	        return action;
	    }

	    // Opens the WordDictation activity,
	    // or, quits the program.
		private void ProcessVoiceAction(string voiceAction)
	    {
	        if(voiceAction != null)
			{
				if(voiceAction.Equals(VoiceDemoConstants.ACTION_START_DICTATION)) 
				{
					OpenVoiceDictationActivity();
	            }
				else if(voiceAction.Equals(VoiceDemoConstants.ACTION_STOP_VOICEDEMO)) 
				{
					//Log.i("VoiceDemo activity has been terminated upon start.");
					this.Finish();
	            }
				else 
				{
					//Log.w("Unknown voice action: " + voiceAction);
	            }
	        }
			else 
			{
				//Log.w("No voice action provided.");
	        }
	    }

		public void OpenVoiceDictationActivity()
	    {
			Intent intent = new Intent(this, typeof(VoiceDictationActivity));
			StartActivity(intent);
	    }


	    ///////////////////////////////////////////////////////
	    /// Gesture handling
	    //

		public override bool OnGenericMotionEvent(MotionEvent e)
	    {
			if (_gestureDetector != null)
			{
				return _gestureDetector.OnMotionEvent(e);
	        }
			return base.OnGenericMotionEvent(e);
	    }

		private Android.Glass.Touchpad.GestureDetector CreateGestureDetector(Context context)
		{
			Android.Glass.Touchpad.GestureDetector gestureDetector = new Android.Glass.Touchpad.GestureDetector (context);
			//Create a base listener for generic gestures
			gestureDetector.SetBaseListener (new MyGestureDetector (this));

			return gestureDetector;
		}

		private class MyGestureDetector : Android.Glass.Touchpad.GestureDetector.IBaseListener
		{
			VoiceDemoActivity _context;

			public MyGestureDetector (VoiceDemoActivity context)
			{
				_context = context;
			}

			#region IBaseListener implementation
			public bool OnGesture (Android.Glass.Touchpad.Gesture gesture)
			{
				if (gesture == Android.Glass.Touchpad.Gesture.Tap) 
				{
					_context.OpenVoiceDictationActivity();
					return true;
				} // etc...
				return false;
			}
			#endregion
			#region IDisposable implementation
			public void Dispose ()
			{
			}
			#endregion
			#region IJavaObject implementation
			public IntPtr Handle {
				get {
					return new IntPtr ();
				}
			}
			#endregion
		}

	}
}
